### ISSUE TITLE
Feature Request: Show function return value in the debugger

### ISSUE THREAD

**Issue by Gruntfuggly** [microsoft/vscode-cpptools#3214]
Would it be possible to show the last function return value somewhere in the debugger window - preferrably variables?

I know it's possible to get at using gdb commands, but it's fairly convoluted. Not sure how difficult it is on Windows though?

**Comment by natrist** [comment]
I also think you should be able to see the return value of a function by hovering it, just like in Visual Studio.

**Comment by HorstBaerbel** [comment]
Imo this is pretty basic and pretty important. I wonder why so few people miss this...

**Comment by dreamer2q** [comment]
Why not implement this feature since gdb supports to inspector the return value from a function.

**Comment by Trass3r** [comment]
Should be shown in the variables view after exiting or stepping over a function, just like in VS.
Requires DAP support I guess, @weinand?

**Comment by weinand** [comment]
@Trass3r no, VS Code's new JavaScript debugger shows return values without DAP support:

<img width="775" alt="2020-07-22_21-25-52" src="https://user-images.githubusercontent.com/1898161/88219572-176b7100-cc62-11ea-8eee-b762609186e3.png">

**Comment by Trass3r** [comment]
Apparently [this](https://github.com/microsoft/vscode-js-debug/commit/fd6f258ec9968d61639a082ed02e49b0f249fe44) commit.
But a bit weird to see the return value while being inside the function rather than after returning.

**Comment by weinand** [comment]
@connor4312 VS Code should not use private protocol. If DAP is not sufficient to support showing return values, please create a DAP feature request. 

@Trass3r you are invited to file a feature request for whatever you want to add to the DAP.

**Comment by connor4312** [comment]
@weinand js-debug doesn't use anything private/special in DAP. We just return a variable that we name "Return value."

> But a bit weird to see the return value while being inside the function rather than after returning.

js-debug will show the return value whenever it's present in the information we get from the runtime. This happens right after you step over a "return" statement, before the function is popped from the stack (though if you're running with sourcemaps, this location can be obscured; tooling seems to be somewhat inconsistent in mapping the whitespace range after return statements)

---

I agree that this is something that can and should be handled adapter side. We do some fairly advanced things with the return values in js-debug (e.g. allowing them to be used in watch expressions and the repl) and I never ran into a limitation imposed by the DAP.

**Comment by weinand** [comment]
@connor4312 thanks for your confirmation that you are not using private protocol! :-)

**Comment by Trass3r** [comment]
Interesting read: https://stackoverflow.com/questions/267674/how-to-inspect-the-return-value-of-a-function-in-gdb
So in theory instead of stepping over a call you could do something like stepping into it and then execute finish. Or a platform-dependent approach like inspecting RAX etc.

**Comment by Trass3r** [comment]
Indeed the stopped event after stepping out contains the return value:
```
->*stopped,reason="function-finished",frame={addr="0x00007ffff7cef3de",func="myouterfunc",args=[{name="this",value="0x5555556acd20"}],file="file.cpp",fullname="file.cpp",line="67"},gdb-result-var="$1",return-value="(ReturnType &) @0x5555556b2a10: {............}",thread-id="1",stopped-threads="all",core="2"
```

https://github.com/microsoft/MIEngine/blob/54b7d6b6dcc11d99297bcad8df765870f4e80da0/src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs#L1238

**Comment by Trass3r** [comment]
I got a prototype of the functionality when stepping out. It is indeed a pure adapter change.

**Comment by Gruntfuggly** [comment]
Not sure when this changed, but I now see 
```
$ReturnValue: <...>
```
in the local variables.

Thank you!

**Comment by Trass3r** [comment]
Yeah fixed in https://github.com/microsoft/MIEngine/pull/1036, but only when stepping out, not over.

https://sourceware.org/bugzilla/show_bug.cgi?id=26354

**Comment by gaowujie2004** [comment]
Hello, what if it's a native function? Vscode debugging cannot go inside native functions.

Example:
document.querySelect('html')
